### PR TITLE
Feature/implement tiovx pad

### PR DESCRIPTION
**Pull request by emmadrigal**

### CODE REVIEW

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+GST_DEBUG_CATEGORY_STATIC (gst_tiovx_pad_debug_category);
+#define GST_CAT_DEFAULT gst_tiovx_pad_debug_category
+
+struct _GstTIOVXPad
+{
+  GstPad base;
+
+  GstTIOVXBufferPool *buffer_pool;
+
+    gboolean (*notify_function) (GstElement * notify_element);
```
Comment (dchvs) [inline gst-libs/gst/tiovx/gsttiovxpad.c:86]: Identation

Comment (emmadrigal) [reply gst-libs/gst/tiovx/gsttiovxpad.c:86]: This is done by the automatic tool, moving it to match the other members causes it to go back to this indentation

File: tests/check/libs/gsttiovxpad.c [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+    gboolean sink_pad)
+{
+  GstPadDirection direction;
+  vx_status status;
+
+  if (sink_pad) {
+    direction = GST_PAD_SINK;
+  } else {
+    direction = GST_PAD_SRC;
+
```
Comment (dchvs) [inline tests/check/libs/gsttiovxpad.c:138]: Extra space.

File: tests/check/libs/gsttiovxpad.c [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+}
+
+static void
+init (GstTIOVXPad ** pad, vx_context * context, vx_reference * reference,
+    gboolean sink_pad)
+{
+  GstPadDirection direction;
+  vx_status status;
+
+  if (sink_pad) {
```
Comment (dchvs) [inline tests/check/libs/gsttiovxpad.c:134]: You can compact this to one line:

`direction = sink_pad ? GST_PAD_SINK : GST_PAD_SRC`

File: tests/check/libs/gsttiovxpad.c [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+  caps = gst_caps_new_simple ("video/x-raw",
+      "format", G_TYPE_STRING, kGstImageFormat,
+      "width", G_TYPE_INT, kImageWidth,
+      "height", G_TYPE_INT, kImageHeight, NULL);
+  query = gst_query_new_allocation (caps, TRUE);
+
+  ret = gst_pad_query (GST_PAD (pad), query);
+  fail_if (!ret, "Unable to query pad");
+
+  for (npool = 0; npool < gst_query_get_n_allocation_pools (query); ++npool) {
```
Comment (dchvs) [inline tests/check/libs/gsttiovxpad.c:185]: You're the expert with the Gst pools, but I got the question, why do we start at `n=1`, with the statement `++npool`?

File: tests/check/libs/gsttiovxpad.c [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+}
+
+static void
+initialize_tiovx_buffer_pool (GstBufferPool ** buffer_pool)
+{
+  GstStructure *conf = NULL;
+  GstCaps *caps = NULL;
+  GstVideoInfo info;
+  gboolean ret = FALSE;
+  vx_context context;
```
Comment (dchvs) [inline tests/check/libs/gsttiovxpad.c:207]: These are pointers, initialize them all.

**Review by dchvs — APPROVED** [review]

File: gst-libs/gst/tiovx/gsttiovxmeta.c [hunk 6]
```diff
@@ -204,10 +204,12 @@ gst_buffer_add_tiovx_meta (GstBuffer * buffer, const vx_reference exemplar,
       gst_tiovx_meta_get_info (), NULL);
   tiovx_meta->array = array;
 
+  tiovx_meta->image_info.data = (guint8 *) mem_start;
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxmeta.c:207]: Why do we need access to the data? This feels dangerous.

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+  self->notify_function = notify_function;
+  self->notify_element = element;
+}
+
+void
+gst_tiovx_pad_install_chain (GstTIOVXPad * pad,
+    gboolean (*chain_function) (GstElement * element, GstBuffer * buffer),
+    GstElement * element)
+{
+  GstTIOVXPad *self = GST_TIOVX_PAD (pad);
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxpad.c:246]: pointer guards for all methods please.

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+  GstMemory *in_memory = NULL;
+  GstCaps *in_caps = NULL;
+  GstStructure *in_config = NULL;
+  GstFlowReturn flow_return = GST_FLOW_ERROR;
+  GstVideoInfo caps_info;
+  GstMapInfo in_info;
+  gboolean ret = FALSE;
+  gint i = 0;
+
+  g_return_val_if_fail (pool, NULL);
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxpad.c:360]: pointer guards

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+
+  ret = gst_memory_map (in_memory, &in_info, GST_MAP_READ);
+  if (FALSE == ret) {
+    GST_ERROR_OBJECT (pad, "Unable map input memory");
+    goto free_inmem;
+  }
+
+  if (!gst_buffer_pool_is_active (GST_BUFFER_POOL (pad->buffer_pool))) {
+    gst_buffer_pool_set_active (GST_BUFFER_POOL (pad->buffer_pool), TRUE);
+  }
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxpad.c:391]: This is very random here. Not the place for this.

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+  if (NULL == meta) {
+    GST_ERROR_OBJECT (pad, "Unable to acquire TIOVX meta");
+    goto unmap_inmem;
+  }
+
+  for (i = 0; i < caps_info.finfo->n_planes; i++) {
+    memmove (meta->image_info.data + meta->image_info.plane_offset[i],
+        in_info.data + caps_info.finfo->poffset[i],
+        meta->image_info.plane_sizes[i]);
+  }
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxpad.c:413]: Don't do this manually. Just use gst_buffer_copy or gst_buffer_copy_into (can't recall the difference)

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+GstFlowReturn
+gst_tiovx_pad_acquire_buffer (GstTIOVXPad * pad, GstBuffer ** buffer,
+    GstBufferPoolAcquireParams * params)
+{
+  g_return_val_if_fail (pad, GST_FLOW_ERROR);
+  g_return_val_if_fail (buffer, GST_FLOW_ERROR);
+
+  if (!gst_buffer_pool_is_active (GST_BUFFER_POOL (pad->buffer_pool))) {
+    gst_buffer_pool_set_active (GST_BUFFER_POOL (pad->buffer_pool), TRUE);
+  }
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxpad.c:472]: normally this is done right after creating and configuring the pool. No need to place it everywhere.

File: gst-libs/gst/tiovx/gsttiovxpad.c [hunk 12]
```diff
@@ -0,0 +1,482 @@
...
+
+  return gst_buffer_pool_acquire_buffer (GST_BUFFER_POOL (pad->buffer_pool),
+      buffer, params);
+}
+
+static void
+gst_tiovx_pad_finalize (GObject * object)
+{
+  G_OBJECT_CLASS (gst_tiovx_pad_parent_class)->finalize (object);
```
Comment (michaelgruner) [inline gst-libs/gst/tiovx/gsttiovxpad.c:481]: Do you need to free the pool?

**Review by michaelgruner — CHANGES_REQUESTED** [review]

File: tests/check/libs/gsttiovxpad.c (hunk 4 above)
Comment (emmadrigal) [reply tests/check/libs/gsttiovxpad.c:185]: We don't start at 1, the `++npool` is called at the end of the for loop

The difference between `++npool` and `npool++` is relevant for cases like this:
```
 npool = 0;
 i = npool++
```
and 
```
 npool = 0;
 i = ++npool
```

In the first case `i = 0` and in the second one `i = 1`.

File: tests/check/libs/gsttiovxpad.c (hunk 3 above)
Comment (emmadrigal) [reply tests/check/libs/gsttiovxpad.c:134]: I've always found that gramatic not very clear. I prefer to be more explicit

**Pull request merged**